### ISSUE TITLE
add support for completion resolve

### ISSUE THREAD

**Issue by prabirshrestha** [prabirshrestha/vim-lsp#304]
Need to wait for https://github.com/neovim/neovim/pull/9616 to be merged in neovim first. Also should port it to vim.

**Comment by mattn** [comment]
Is it same feature of TextChangedP ?

**Comment by chemzqm** [comment]
> Is it same feature of TextChangedP ?

No, TextChangedP fires on text change when pum visible, this autocmd fires on pum redraw, TextChangedP could be fired or not.

**Comment by mattn** [comment]
What is the trigger? When item is selected? If I understand the details of the event, I can port it to vim. :)

(I'm sorry. I am careful not to read the source of neovim intentionally as possible for avoiding doubt about code copyright since I am mainly working on vim always)

**Comment by chemzqm** [comment]
> What is the trigger? When item is selected?

It triggers on each popup menu drawn (`ins_compl_show_pum` function in `edit.c`), `v:event.completeitem` is `{}` when no complete item is selected.

**Comment by mattn** [comment]
Thank you. I'll try to implement it in later.

**Comment by mattn** [comment]
Ah, if you have plan to send PR to Vim, please do! :)

**Comment by prabirshrestha** [comment]
LSP supports completion resolve which can be used by editors to ask for more information about the selected item.

Here is an example of VsCode which even shows a callout near the selected item. This can allow the lang servers to be fast and return only the result and only process function signature details when an item is selection. Also reduces the chattiness when taking with server over stdio.

![image](https://user-images.githubusercontent.com/287744/53151710-b9b49000-3568-11e9-9e95-8c5215a92a82.png)

I have added comments in https://github.com/neovim/neovim/pull/9616#issuecomment-465892624 on my thoughts. Pasting it here for reference.

-------------

`PumRender` doesn't seem like a good name. Just by looking at the name I don't know what it does. I also agree with roxma resize shouldn't call `CompleteSelectionChanged`. `PumRedraw` might be a better name when pum gets updates or redrawn. But I think this is something we shouldn't expose yet.

What about this? Add a `autocmd` that notifies selection changed by `CompleteSelectionChanged` or whatever name we come up with. Then add a function `pumselected()` which can be called anywhere similar to `pumvisible()`. `pumselected()` can return `v:null` if nothing is selected or if pum is not visible.

```vim
function! s:callback() abort
    let selected = pumselected()
endfunction

autocmd CompleteSelectionChanged, VimResize call s:callback()
```

`pumselected()` also make it easy to debounce selection when the user is changing selections fast.

```vim
function! s:pum_selection_changed_debounce() abort
    if exists('s:pum_selection_changed_timer')
        call timer_stop(s:pum_selection_changed_timer)
        unlet s:pum_selection_changed_timer
    endif
    let s:pum_selection_changed_timer = timer_start(100, function! s:pum_selection_changed)
endfunction

function! s:pum_selection_changed(...) abort
    " one could also check for pumvisible() && pumselected()
    let selected = pumselected()
    if selected
        " do something with selected['user_data']
    endif
endfunction

autocmd CompleteSelectionChanged call s:pum_selection_changed_debounce()
```

`CompleteDone` can still continue to use `v:completed_item` and `pumselected()` should return `v:null` since pum is already closed.

**Comment by leighmcculloch** [comment]
In vim8 is there already an event we can hook this into? Or is this likely to be a NeoVim only feature?

To add to @prabirshrestha's comments above about how resolve can be used, it can pull the details like function parameters, but some language servers will also provide the `documentation` parameter then too. (e.g. the [java-language-server](https://github.com/georgewfraser/java-language-server) does this)

**Comment by prabirshrestha** [comment]
Currently it is a neovim only feature. Requires `MenuPopupChanged` at minimum to implement it and showing in preview window would be flickering the screen. To be better we need floating window support which currently is also only enabled in neovim.

![iohyj8skafm21](https://user-images.githubusercontent.com/287744/54478661-fbd09a80-47d1-11e9-8999-9835a5d80648.gif)
Source: https://www.reddit.com/r/neovim/comments/b1pctc/float_window_support_with_cocnvim/

Having said that vim-lsp is meant for both vim and neovim, so if there is something one has but the other doesn't I'm totally ok including it if it helps the user. We already do that currently for example virtual text support is neovim only at the moment. https://github.com/prabirshrestha/vim-lsp/pull/336

**Comment by leighmcculloch** [comment]
Yup that makes sense. It'd be great to also get this feature in Vim. I see you asked here https://github.com/neovim/neovim/pull/9616#issuecomment-473559486. Is the hope that the author of that patch will submit it to vim as well?

**Comment by chemzqm** [comment]
It's a simple patch.  It doesn't makes too much sense for vim since it doesn't support floating window, the completion resolve request can be send on complete done.

**Comment by leighmcculloch** [comment]
Oh interesting. I think that'd work for some things, but maybe not in some languages like Java where you'll have 10 methods all with the same name but different signatures.

Is there an event that triggers when complete is done though? I don't see one here http://vimdoc.sourceforge.net/htmldoc/autocmd.html#MenuPopup.

**Comment by chemzqm** [comment]
`:h CompleteDone`

**Comment by clason** [comment]
Bump?

**Comment by clason** [comment]
@thomasfaingnaert Independent of how to style the preview using popup/floats, I think this needs to (and can) be addressed first to populate the `info` field.

**Comment by thomasfaingnaert** [comment]
@clason I'll check it out.

**Comment by thomasfaingnaert** [comment]
@clason Does texlab implement it?

**Comment by clason** [comment]
I think so: 
```
Tue 10 Sep 2019 04:27:41 PM CEST:["<---", 5, "texlab", {"response": {"id": 1, "jsonrpc":   "2.0", "result": {"capabilities": {..., "completionProvider": {"resolveProvider": true, ...
```
(also, works in VS Code).

**Comment by clason** [comment]
@prabirshrestha I found another issue preventing from showing the `info` field for texlab responses that is fixed by #500. I now get the expected preview at least, so "all" that's left is the styling.
